### PR TITLE
gateway: make API commands configurable

### PR DESCRIPTION
**Pull request by lgierth**
This is for Delegated Routing in js-libp2p (ipfs/js-ipfs#1459 & libp2p/js-libp2p#120) and supersedes #4595.

It lets us configure the commands available through :8080/api/v0/ -- for delegated routing, we would then e.g. expose `swarm/connect`, `refs`, `dht/findprovs`, and `dht/findpeer`.

By default, `Gateway.APICommands` is empty, which means go-ipfs will use the default :8080 commands.

```sh
> ipfs init
> ipfs &
> curl localhost:5001/api/v0/config?arg=Gateway.APICommands | jq -r .Value
[]
> curl localhost:8080/api/v0/commands | jq -r .Subcommands[].Name
dns
version
commands
cat
get
dag
refs
block
ls
resolve
name
object
> killall ipfs

> ipfs config Gateway.APICommands --json '["config"]'
> ipfs daemon &
> curl localhost:8080/api/v0/commands | jq -r .Subcommands[].Name
config
commands
> killall ipfs
```

### CODE REVIEW

**Comment by lgierth** [comment]
Whoops, totally forgot updating the go tests. otherwise still happy about review :)

**Comment by lgierth** [comment]
I am pretty sure that sharness t0236-cli-api-dns-resolve.sh is failing on Jenkins because of a `nc` or `ipfs` process leaking on the worker.

**Comment by lgierth** [comment]
This one is ready for review

File: core/commands/root.go [hunk 1]
```diff
@@ -147,64 +148,46 @@ var rootSubcommands = map[string]*cmds.Command{
...
+		out := subset
+		for _, elem := range pathelems {
+			nextIn, ok := in.Subcommands[elem]
+			if !ok {
+				return nil, fmt.Errorf("unknown command: %s", path)
+			}
+
+			nextOut := new(cmds.Command)
+			*nextOut = *nextIn
+			nextOut.Subcommands = map[string]*cmds.Command{}
```
[the remark this answers is not in the document]
Comment (keks) [reply core/commands/root.go]: I updated the tests and now they fail. Could you fix this?

**Comment by lidel** [comment]
@lgierth did this ever land in some other PR? was there anything more to the story?

We were trying to use `Gateway.APICommands` in context of https://github.com/ipfs/js-ipfs/issues/2155#issuecomment-505005256 but it did not work, and @mburns  had to redirect to API port.

**Comment by Stebalien** [comment]
We never did.

**Pull request closed without merging**